### PR TITLE
typing: fix App's core API - api

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

fixes App's core API - api

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda @justusschock @awaelchli

### CODE REVIEW

File: src/lightning/app/core/api.py [hunk 1]
```diff
@@ -490,7 +493,7 @@ def start_server(
         else:
             logger.info(f"Your app has started. View it in your browser: http://{host}:{port}/view")
         if has_started_queue:
-            LightningUvicornServer.has_started_queue = has_started_queue
+            LightningUvicornServer.has_started_queue = has_started_queue  # type: ignore[assignment]
```
Comment (awaelchli) [inline src/lightning/app/core/api.py:496]: You can type it as Optional[bool]

File: src/lightning/app/core/api.py [hunk 2]
```diff
@@ -410,21 +413,21 @@ async def websocket_endpoint(websocket: WebSocket):
...
+async def api_catch_all(request: Request, full_path: str) -> None:
     raise HTTPException(status_code=404, detail="Not found")
 
 
 # Serve frontend from a static directory using FastAPI
 fastapi_service.mount("/static", StaticFiles(directory=frontend_static_dir, check_dir=False), name="static")
 
 
-async def frontend_route(request: Request, full_path: str):
+async def frontend_route(request: Request, full_path: str):  # type: ignore[no-untyped-def]
```
Comment (awaelchli) [inline src/lightning/app/core/api.py:424]: this looks like Union[str, templates.TemplateResponse]

File: src/lightning/app/core/api.py [hunk 3]
```diff
@@ -210,13 +211,13 @@ async def get_state(
...
     with lock:
-        x_lightning_session_uuid = TEST_SESSION_UUID
+        x_lightning_session_uuid = TEST_SESSION_UUID  # type: ignore[assignment]
         state = global_app_state_store.get_app_state(x_lightning_session_uuid)
         global_app_state_store.set_served_state(x_lightning_session_uuid, state)
         return state
 
 
-def _get_component_by_name(component_name: str, state):
+def _get_component_by_name(component_name: str, state: dict):  # type: ignore[no-untyped-def]
```
Comment (awaelchli) [inline src/lightning/app/core/api.py:220]:
```suggestion
def _get_component_by_name(component_name: str, state: dict) -> Union[LightningFlow, LightningWork]
```

**Review by awaelchli — APPROVED** [review]

**Review by ethanwharris — APPROVED** [review]

File: src/lightning/app/core/api.py (hunk 2 above)
Comment (Borda) [reply src/lightning/app/core/api.py:424]: tried and do not like it because of the template

File: src/lightning/app/core/api.py (hunk 3 above)
Comment (Borda) [reply src/lightning/app/core/api.py:220]: there is cyclic import :/

**Comment by Borda** [comment]
@awaelchli @ethanwharris any idea why e2e apps are failing?

**Pull request merged**